### ISSUE TITLE
Meta-issue for making admin-generated content translatable

### ISSUE THREAD

**Issue by mlovic** [AyuntamientoMadrid/consul#1612]
This is a _meta-issue_ for all changes related to making admin-generated content translatable.

## Relevant issues and PRs

Original implementation of the shared "Translatable" code + application of this code to Milestones:
  - [Allow admin generated content to be translatable](https://github.com/AyuntamientoMadrid/consul/pull/1435) (merged)

More issues to apply this general "translatable" code to other resources (I will refer to these as **feature PRs**):

 - <del>[Banners](https://github.com/consul/consul/issues/2738) (merged PR https://github.com/AyuntamientoMadrid/consul/pull/1585)
 - [Participatory budgets](https://github.com/consul/consul/issues/2737#issuecomment-413580923) (put on hold due to incompatibility with current Javascript implementation)
 - <del>[Admin notifications](https://github.com/consul/consul/issues/2742)</del> (merged PR https://github.com/AyuntamientoMadrid/consul/pull/1596)
 - <del>[Collaborative legislation](https://github.com/consul/consul/issues/2736)</del> (merged PR https://github.com/AyuntamientoMadrid/consul/pull/1614)
 - <del>[Homepage content](https://github.com/consul/consul/issues/2668)</del> (merged PR https://github.com/AyuntamientoMadrid/consul/pull/1588)
 - <del>[Polls](https://github.com/consul/consul/issues/2735)</del> (merged PR by a different contributor https://github.com/consul/consul/pull/2833)
 - <del>[Custom pages](https://github.com/consul/consul/issues/2741)</del> (merged PR by a different contributor https://github.com/consul/consul/pull/2839)

Additionally, there are 3 more open PRs with changes to the code that all "translatable" resources share. I will call these **refactor PRs**, to distinguish them from the above _feature PRs_. These changes include the corresponding changes to the resources that are already in master.

  - <del>[Change Translatable implementation to accommodate new requirements](https://github.com/AyuntamientoMadrid/consul/pull/1602#issuecomment-412954579)</del> (merged)
  - <del>[Refactor code shared by admin-translatable resources](https://github.com/AyuntamientoMadrid/consul/pull/1613)</del> (merged)
 - <del>[Refactor translatable specs](https://github.com/AyuntamientoMadrid/consul/pull/1617)</del> (merged)

## Review and Merging

Because a lot of this work has been done this August while master was frozen, there is a lot of unmerged code that depends on the _refactor PRs_.

Instead of waiting for the refactor PRs to be in master to resolve conflicts, I've chosen to incorporate the refactors into the unmerged feature branches (more discussion at https://github.com/AyuntamientoMadrid/consul/pull/1602#issuecomment-412954579).

Here is what the commit graph looks like right now (each ref corresponds to an open PR):

![selection_123](https://user-images.githubusercontent.com/7111622/44848447-0fae1800-ac57-11e8-84e1-cf76e1dc3b99.png)

Starting from master, there are 3 refactor branches (one following from the other), and then the feature PRs follow in parallel from there.

The idea is to first merge the two refactor PRs (in commit order), then the feature PRs (in any order). Each feature branch can be rebased onto master before merging to keep the commit history on master as linear as possible.

Note that this only applies to the feature PRs that I've opened. There are 2 more by other contributors.

## Additional notes

 - These changes don't include data migrations for the translatable attributes, or schema migrations to drop the old columns. _For example: for banners, there is a migration to create a table `banner_translations`  with the column `title`, but there is no code to 1) create translation records for the default locale, 2) copy the data in `banners.title` to `banner_translations.title`, and 3) remove `banners.title` column._
 - The backport to consul/consul should probably not include 16cc6e6

**Comment by mlovic** [comment]
> @mlovic Thank you so much for the explanation heart_eyes! Just one question: do you know whether consul#2839 depends on any of these PRs too?

Yes it does, I didn't even know about that issue/PR. I'll add it to this issue and leave a comment on it. 

> BTW, there's a small typo in the TL; DR section: it's 1588 instead of 158.

Thanks, fixed!

**Comment by voodoorai2000** [comment]
This is very helpful, appreciated it @mlovic 👌

**Comment by javierm** [comment]
Closing this one; since there's only one issue left (budgets), we don't need a meta-issue anymore.

(Although technically pull request #1669 is certainly related since it changes code affecting all translatable models).

Thank you sooooo much for your work on this one @mlovic :heart_eyes:.
